### ISSUE TITLE
Preprocess code for SEED-V and MoBI?

### ISSUE THREAD

**Issue by dyhan316** [935963004/LaBraM#48]
Hello, I was wondering if you have the code for preprocessing SEED-V and MoBI for use in fine-tuning?

**Comment by ShogoNoguchi** [comment]
Hi! Sorry, I'm not the author, But
I'm preparing it(for SEEDV)! Please wait for about a week.

**Comment by dyhan316** [comment]
No need to be sorry:) glad to meet someone who also is trying to replicate the paper!

Do you have one for other datasets? (TUEG and such)

Also it seems that for one of their datasets they use a bipolar reference scheme. Did you notice this too?

**Comment by ShogoNoguchi** [comment]
Thank you for your reply!

Sorry, I only have SEEDV.

I think you're right. In make_TUEV.py, uncomment #signals = convert_signals(signals, Rawdata) and set ”ch_names” to the following in the branch of the get_dataset function（in run_class_fintuning.py） to make it work properly.

bipolar_ch_names = [
"FP1-F7", "F7-T7", "T7-P7", "P7-O1",
"FP2-F8", "F8-T8", "T8-P8", "P8-O2",
"FP1-F3", "F3-C3", "C3-P3", "P3-O1",
"FP2-F4", "F4-C4", "C4-P4", "P4-O2"
]

All of these channels are included in the following list in utils.py, so the spatial embedding will be added correctly.

standard_1020 = [ 'FP1', 'FPZ', 'FP2', 'AF9', 'AF7', 'AF5', 'AF3', 'AF1', 'AFZ', 'AF2', 'AF4', 'AF6', 'AF8', 'AF10', \ 'F9', 'F7', 'F5', 'F3', 'F1', 'FZ', 'F2', 'F4', 'F6', 'F8', 'F10', \ 'FT9', 'FT7', 'FC5', 'FC3', 'FC1', 'FCZ', 'FC2', 'FC4', 'FC6', 'FT8', 'FT10', \ 'T9', 'T7', 'C5', 'C3', 'C1', 'CZ', 'C2', 'C4', 'C6', 'T8', 'T10', \ 'TP9', 'TP7', 'CP5', 'CP3', 'CP1', 'CPZ', 'CP2', 'CP4', 'CP6', 'TP8', 'TP10', \ 'P9', 'P7', 'P5', 'P3', 'P1', 'PZ', 'P2', 'P4', 'P6', 'P8', 'P10', \ 'PO9', 'PO7', 'PO5', 'PO3', 'PO1', 'POZ', 'PO2', 'PO4', 'PO6', 'PO8', 'PO10', \ 'O1', 'OZ', 'O2', 'O9', 'CB1', 'CB2', \ 'IZ', 'O10', 'T3', 'T5', 'T4', 'T6', 'M1', 'M2', 'A1', 'A2', \ 'CFC1', 'CFC2', 'CFC3', 'CFC4', 'CFC5', 'CFC6', 'CFC7', 'CFC8', \ 'CCP1', 'CCP2', 'CCP3', 'CCP4', 'CCP5', 'CCP6', 'CCP7', 'CCP8', \ 'T1', 'T2', 'FTT9h', 'TTP7h', 'TPP9h', 'FTT10h', 'TPP8h', 'TPP10h', \ "FP1-F7", "F7-T7", "T7-P7", "P7-O1", "FP2-F8", "F8-T8", "T8-P8", "P8-O2", "FP1-F3", "F3-C3", "C3-P3", "P3-O1", "FP2-F4", "F4-C4", "C4-P4", "P4-O2" ]

However, since LaBram uses an absolute spatial embedding method, the accuracy of the learning may not be guaranteed.

This is exactly what I'm thinking about right now! Let's think about it together!!

**Comment by dyhan316** [comment]
Yes!!!! It would be wonderful to work together to make this work :) Post any question or suggestions on this GitHub issue and we can work together :)

**Comment by dyhan316** [comment]
I've found that TUEG only has a subset of the channels the channels you mentioned (standard_1020 (not really 1020 haha )) 

What do you mean by "absolute spatial embedding method"?  I thought that LaBraM learns an embedding for each channel separately, and not care about the actual spatial positions of the channels?

**Comment by 935963004** [comment]
Thanks for your question! There is some important information I need to clarify, that is we did not use any bipolar reference in this project, including pre-training datasets and downstream datasets such as TUAB and TUEV, where we directly used channels like "F4-REF".

**Comment by dyhan316** [comment]
Oh!! you're here :) 

Thank you for the response! I must've been mistaken but 

https://github.com/935963004/LaBraM/blob/1151c3ee921fc4ba5787fc923517672efc426257/dataset_maker/make_TUEV.py#L49

I thought that this code shows that (attached image below) you computed the bipolar reference ?

<img width="608" alt="image" src="https://github.com/user-attachments/assets/c6b51530-33d1-47b4-ac0b-ac0722326c80" />

**Comment by dyhan316** [comment]
Also, it since you're here I have a few more questions if you don't mind : 
1. Did you do any QA (dropping bad channels, or clipping too large voltage values) during preprocessing?
2. In TUEG, did you use the LE references or only AR references? (i.e. only channel names "REF F1-REF" or also "REF F1-LE")? 
3. What was the total size of the TUEG dataset you used? Was it close to the original dataset? (1.6TB, with around 60k edf files)
4. In the `standard_1020` list, there seems to be channels that are not included in TUEG. 1. Did you train those channel's latent embeddings using different datasets? 2. Also, what about channels like "FP2-F4"?

**Comment by 935963004** [comment]
Yeah, this is actually the function for bipolar reference. We preserve it but do not call it, as you can see in line 157 where we comment it.

**Comment by dyhan316** [comment]
Oh I see :) that clears it up a bit!

**Comment by dyhan316** [comment]
> Also, it since you're here I have a few more questions if you don't mind :
> 
> 1. Did you do any QA (dropping bad channels, or clipping too large voltage values) during preprocessing?
> 2. In TUEG, did you use the LE references or only AR references? (i.e. only channel names "REF F1-REF" or also "REF F1-LE")?
> 3. What was the total size of the TUEG dataset you used? Was it close to the original dataset? (1.6TB, with around 60k edf files)
> 4. In the `standard_1020` list, there seems to be channels that are not included in TUEG. 1. Did you train those channel's latent embeddings using different datasets? 2. Also, what about channels like "FP2-F4"?

(more questions)
5. Could you share the code for running fine-tuning on MOBI and SEED? It'd be greatly appreciated!

**Comment by 935963004** [comment]
For your question:
1. No, we didn't. However, this is definitely a meaningful direction to filter bad-quality data. It is worthwhile to try!
2. We only use AR references and skip LE-reference files.
3. We only use four subsets of TUEG as mentioned in the Appendix.
4. 1. Yes, we use many other datasets to jointly optimize the channel's embeddings. 2. We did not use bipolar references because it is not common in EEG.
5. I'm sorry I can't. There are some inner reasons for that. This work was done during my internship in a company. Part of the codes for these two datasets and the access to the dataset was not done by me. However, we have provided all the settings and hyperparameters we used for the two datasets, I think they can be fully replicated.

**Comment by dyhan316** [comment]
Thank you so much for your answer :) @ShogoNoguchi  and I can probably work on that :) thanx again for the awesome paper :)

**Comment by ShogoNoguchi** [comment]
dyhan,
The author's explanation was helpful to me as well.

Thank you for the exciting suggestion! If I have any questions or suggestions, I will put them here so that I can discuss them with you from now on!!

As the author explained, I completely agree with you on what I meant by "absolute spatial embedding method".

# As stated immediately after equation (2) on page 4 of the paper ("temporal and spatial embeddings act as absolute position encoding."),

I think that this embedding only has the meaning of identifying the electrodes and does not take into account the positional information of the electrodes.

Thank you for providing deep insight. This Issue is really helpful to me.

If you are interested, please take a look at my Issue #48.

Thank you!

**Comment by ShogoNoguchi** [comment]
In fact, I am aiming to apply this technology to classify emotions using "Muse2".
With Muse2, we can obtain data for TP9, AF7, AF8, and TP10 of the eeg 10-10 system, but because it is a simple device, I must take into account the possibility of electrode position shifting.

Therefore, I am planning to achieve this by using simulation data that takes into account the possibility of electrode position shifting. I am a beginner especially when it comes to EEG, so I would appreciate any knowledge you can share with me!!

Best

**Comment by dyhan316** [comment]
Oh I already read your comments before you deleted it hehe 

If you want, we can have a meeting. I can also suggest some papers that do spatial embedding that you said if you want :)

**Comment by ShogoNoguchi** [comment]
Thank you so much!! I appreciate your thoughtfulness!

**Comment by dyhan316** [comment]
@ShogoNoguchi Hey, have you been able to replicate SEED? I was thinking of replicating MOBI

**Comment by ShogoNoguchi** [comment]
I'm making good progress with finetune now! Wow, you're trying to replicate the results from MOBI, so I'm sure you'll gain some new insights from that. I'm ooking forward to it!

**Comment by INGg** [comment]
> I'm making good progress with finetune now! Wow, you're trying to replicate the results from MOBI, so I'm sure you'll gain some new insights from that. I'm looking forward to it!

Could you please share the code used to process SEED data in this project? I would greatly appreciate it!

**Comment by TousenKaname** [comment]
> I'm making good progress with finetune now! Wow, you're trying to replicate the results from MOBI, so I'm sure you'll gain some new insights from that. I'm looking forward to it!

Hi, I'm algo trying to replicate the resuilts from MOBI but MOBI dataset is regression task. The code of the model only contains classification task code. Should we write code for MOBI specifically?

**Comment by nimingwu** [comment]
Hi @dyhan316 , I saw you mentioned you were thinking of replicating MoBI. Were you able to get it working? I'm currently trying and stuck at test R² ≈ 0.12 with LaBraM-Base (paper reports 0.288). I've followed the preprocessing exactly as described in Appendix F (100→200 Hz, 0.1-75 Hz bandpass, 50 Hz notch, removed EOG, 2s windows / 50ms stride, 10min/5min/5min split, ÷90 normalization, all 8 subjects). Fine-tuning hyperparameters are identical to the paper's Table 5. Tried both center and end-of-window alignment for joint labels, tried both scipy and MNE filtering — results barely changed. Would really appreciate any tips if you managed to get closer to the reported numbers, or if you'd be willing to compare notes.
